**Numpy exercises**

This is a collection of exercises that have been collected in the numpy mailing list, on stack overflow and in the numpy documentation. The goal of this collection is to offer a quick reference for both old and new users but also to provide a set of exercises for those who teach.

If you find an error or think you've a better way to solve some of them, feel free to open an issue at https://github.com/rougier/numpy-100. File automatically generated. See the documentation to update questions/answers/hints programmatically.

**1. Import the numpy package under the name np**

In [70]:
import numpy as np #importing numpy.
arr=np.array([1,3,5,7,9]) #create arr array.

In [71]:
import numpy as np

In [10]:
print(np.__version__)
np.show_config()

1.23.5
blas_armpl_info:
  NOT AVAILABLE
blas_mkl_info:
  NOT AVAILABLE
blis_info:
  NOT AVAILABLE
openblas_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/Users/brunasantos/anaconda3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
blas_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/Users/brunasantos/anaconda3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
lapack_armpl_info:
  NOT AVAILABLE
lapack_mkl_info:
  NOT AVAILABLE
openblas_lapack_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/Users/brunasantos/anaconda3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
lapack_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/Users/brunasantos/anaconda3/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
Supported SIMD extensions in this NumPy install:
    baseline = SSE,SSE2,SSE3
    found = SSSE3,SSE41,POPCNT,SSE42,AVX,F16C,FMA3,AVX2
   

**2. Create a null vector of size 10**

In [72]:
np.zeros(10)

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

**3. Create a null vector of size 10 but the fifth value which is 1**

In [110]:
v = np.array([int(x==4) 
 for x in range(10)])
print(v)

[0 0 0 0 1 0 0 0 0 0]


**4. Create a vector with values ranging from 10 to 49**

In [91]:
np.arange(10,50)

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
       27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43,
       44, 45, 46, 47, 48, 49])

**5. Create a 3x3 matrix with values ranging from 0 to 8**

In [96]:
x = np.arange(9).reshape(3,3)
print (x)

[[0 1 2]
 [3 4 5]
 [6 7 8]]


**6. Find indices of non-zero elements from [1,2,0,0,4,0]**

hint: np.nonzero

In [100]:
v = np.nonzero([1,2,0,0,4,0])
print(v)

(array([0, 1, 4]),)


**7. Create a 3x3 identity matrix**

hint: np.eye

In [101]:
x = np.eye(3)
print (x)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


**8. Create a 3x3x3 array with random values**

hint: np.random.random

In [80]:
x = np.random.random((3,3,3))
print(x)

[[[0.6963047  0.54824165 0.63003907]
  [0.86324683 0.47215946 0.49059768]
  [0.66496851 0.33345488 0.90019896]]

 [[0.90973321 0.06237005 0.57632169]
  [0.49577828 0.55718443 0.73844922]
  [0.32822181 0.34502874 0.54088709]]

 [[0.19963284 0.01002041 0.14578662]
  [0.30840296 0.36019468 0.53854736]
  [0.04637129 0.38023246 0.02653604]]]


**9. Create a 10x10 array with random values and find the minimum and maximum values**

hint: min, max

In [117]:
x = np.random.random((10,10))
print("Original Array:")
print(x) 
xmin, xmax = x.min(), x.max()
print("Minimum and Maximum Values:")
print(xmin, xmax)

Original Array:
[[8.16209530e-01 6.15245072e-01 2.91219334e-01 8.83091027e-01
  1.82573280e-01 2.45784134e-02 6.53733617e-01 7.56302456e-01
  5.46180365e-01 5.52171121e-02]
 [1.91924972e-01 3.24681944e-01 7.79017550e-01 9.52395387e-01
  2.98653333e-01 4.21988019e-01 7.78910023e-01 1.96179717e-01
  8.31066794e-01 2.47617345e-01]
 [2.04933846e-01 8.46847757e-01 6.19360793e-01 6.55675261e-01
  2.27564411e-01 6.29364573e-02 4.66131245e-01 4.61813183e-01
  1.10294582e-01 1.54928493e-01]
 [8.84083724e-01 5.76867183e-01 6.94217189e-01 7.68307201e-01
  2.93860373e-01 1.24332615e-01 3.05871302e-01 8.69742838e-01
  8.29189878e-01 4.75350909e-01]
 [6.92225634e-01 1.96340772e-01 6.76719085e-01 8.54374115e-02
  7.23145041e-02 5.92776217e-01 2.52187930e-01 5.73986564e-01
  4.82802334e-02 6.48452204e-01]
 [6.69124801e-01 7.15983463e-01 7.01431029e-01 6.18072540e-01
  6.61188867e-01 3.38935393e-01 5.58652748e-01 5.54502820e-01
  7.93188510e-02 5.41085649e-02]
 [5.32766997e-01 3.33720632e-01 4.19383023

**10. Create a random vector of size 30 and find the mean value**

hint: mean

In [89]:
v = np.random.random(30)
print(v)

[0.93620049 0.54276777 0.36621354 0.84823095 0.10477989 0.83251192
 0.73849311 0.63012393 0.07332753 0.74338683 0.02526438 0.81730786
 0.9510812  0.93649668 0.31419562 0.40633561 0.87777952 0.50293241
 0.72689506 0.97819955 0.80393338 0.09003219 0.09505138 0.48639698
 0.90240868 0.3822239  0.62769091 0.12579566 0.54281798 0.38520861]


In [90]:
v.mean()

0.5598027843420118

**11. Create a 5x5 matrix with values 1,2,3,4 just below the diagonal**

hint: np.diag

In [102]:
x = np.diag(1+np.arange(4), k = -1)
print (x)

[[0 0 0 0 0]
 [1 0 0 0 0]
 [0 2 0 0 0]
 [0 0 3 0 0]
 [0 0 0 4 0]]


**12. Normalize a 5x5 random matrix**

hint: (x -mean)/std

In [103]:
Z = np.random.random((5,5))
Zmax, Zmin = Z.max(), Z.min()
Z= (Z-Zmin)/(Zmax-Zmin)
print (Z)

[[0.98342712 0.85533341 0.27356972 0.80089971 0.57744462]
 [0.73378635 0.76743841 0.         0.52644104 0.42883136]
 [0.86096071 0.08024195 0.84129307 0.05907241 0.33422692]
 [0.50744258 1.         0.61544688 0.46162287 0.49734414]
 [0.96497547 0.45644905 0.07066212 0.36275465 0.97110868]]


**13. How to find common values between two arrays?**

hint: np.intersect1d

In [113]:
array1 = np.array([0, 10, 20, 40, 60])
print("Array1: ",array1)
array2 = [10, 30, 40]
print("Array2: ",array2)
print("Common values between two arrays:")
print(np.intersect1d(array1, array2))


Array1:  [ 0 10 20 40 60]
Array2:  [10, 30, 40]
Common values between two arrays:
[10 40]


array1 = np.array([0, 10, 20, 40, 60]): Creates a NumPy array with elements 0, 10, 20, 40, and 60.
array2 = [10, 30, 40]: Creates a Python list with elements 10, 30, and 40.
print(np.intersect1d(array1, array2)): The np.intersect1d function returns a sorted array 
    containing the common elements of the two input arrays, ‘array1’ and ‘array2’. 
    
In this case, the common elements are 10 and 40, so the output will be [10 40].

**14. Create a random vector of size 10 and sort it**

hint: sort

In [116]:
v = np.random.random(10)
print("Original array:")
print(x)
x.sort()
print("Sorted array:")
print(x)

Original array:
[[0 0 0 0 0]
 [1 0 0 0 0]
 [0 2 0 0 0]
 [0 0 3 0 0]
 [0 0 0 4 0]]
Sorted array:
[[0 0 0 0 0]
 [0 0 0 0 1]
 [0 0 0 0 2]
 [0 0 0 0 3]
 [0 0 0 0 4]]


x = np.random.random(10): 
    This line creates a 1D array x with 10 random numbers between 0 and 1 using the np.random.random() function. 
    The input integer 10 specifies the output array size.

x.sort(): This line sorts the array x in ascending order using the x.sort() method. 
    The array x is modified in-place, meaning that the sorted values are stored back in the original array.

Finally print() function prints the sorted 1D array x.

**15. Create random vector of size 10 and replace the maximum value by 0**

hint: argmax

In [118]:
Z = np.random.random(10)
Z[Z.argmax()] = 0
print(Z)

[0.14476971 0.438622   0.48767458 0.80848799 0.29307755 0.
 0.56993704 0.53205506 0.27773716 0.35194346]


**16. Subtract the mean of each row of a matrix**

hint: mean(axis=,keepdims=)

In [120]:
X = np.random.rand(5, 10)

# Recent versions of numpy
Y = X - X.mean(axis=1, keepdims=True)

# Older versions of numpy
Y = X - X.mean(axis=1).reshape(-1, 1)
Y

array([[-0.06276938, -0.2741792 ,  0.19996199, -0.19283326, -0.01555784,
        -0.05793655, -0.00979953,  0.38793233,  0.03765188, -0.01247044],
       [-0.1577308 , -0.40429969, -0.29098881, -0.19833719,  0.56300684,
        -0.26291158,  0.07452665,  0.50529845,  0.52490774, -0.35347161],
       [-0.34181914, -0.26975763,  0.13109091, -0.19089166, -0.11024316,
         0.3727739 , -0.37802785,  0.45648356,  0.082347  ,  0.24804406],
       [-0.60748005,  0.26057295,  0.25948713, -0.41204919,  0.36403841,
         0.03138933,  0.03018146,  0.34200552, -0.15706932, -0.11107624],
       [-0.36599977, -0.47227074, -0.14826842, -0.04073174,  0.14626793,
        -0.01988788,  0.14761392,  0.40411455,  0.12221593,  0.22694622]])

**17. How to get the n largest values of an array**

Z = np.arange(10000) np.random.shuffle(Z) n = 5 hint: np.argsort | np.argpartition

In [121]:
Z = np.arange(10000)
print("Original array:")
print(Z)
np.random.shuffle(Z)
n = 5
print (Z[np.argsort(Z)[-n:]])

Original array:
[   0    1    2 ... 9997 9998 9999]
[9995 9996 9997 9998 9999]


Z = np.arange(10000): This statement generates a 1D array of integers from 0 to 9999 (inclusive).

np.random.shuffle(Z): This statement shuffles the elements of the array Z in-place 
    (i.e., the original array is modified) in a random order.

n = 5: This statement sets the variable n to 5, which represents the number of largest elements we want to 
    extract from the shuffled array.

Z[np.argsort(Z)[-n:]]: This statement sorts the indices of the shuffled array Z using np.argsort(z) in ascending order.
    Then, by slicing [-n:], we get the last n indices corresponding to the n largest elements in the array. 
    In this case, since n = 5, the slice returns the index of the largest element in the shuffled array.

print(Z[np.argsort(Z)[-n:]]): Finally print() function prints the largest element in the shuffled array 
    by indexing the array Z with the calculated slice.



**18. Create a random 5*3 matrix and replace items that are larger than 4 by their squares ( Example: 6 --> 36)**

hint: np.where